### PR TITLE
XTB updates

### PR DESCRIPTION
**Pull request by sheepforce**
There is a new major release of the XTB programme. The surrounding infrastructure grew *somewhat* and XTB depends now on a lot of fortran libraries. The Fortran package manager seems to bring a modern build experience and package management to Fortran, finally ... :) However, I am a little bit worried about all the updates these many packages will require. Maybe we should also look into something like r-ryantm's automatic update bot. :sweat_smile:

Also nearly all of them did some cmake mistake and require patching of the generated package-config files. At least the fix is quite easy.

### CODE REVIEW

**Comment by sheepforce** [comment]
At least one test of pysis is failing now. I will collect them and figure out what to do :)

**Comment by ryanswrt** [comment]
I just spent half of yesterday packaging these also, good job; one slight concern on my end was that even though everything compiled; generating the python libs ended up failing the checks with 
```
       > Traceback (most recent call last):
       >   File "/build/python/tblite/library.py", line 28, in <module>
       >     from ._libtblite import ffi, lib
       > ImportError: /build/python/tblite/_libtblite.abi3.so: undefined symbol: __dftd3_model_MOD___vtab_dftd3_model_D3_model
       >
       > During handling of the above exception, another exception occurred:
       >
       > Traceback (most recent call last):
       >   File "/nix/store/xcaaly5shfy227ffs8nipxrd49b56iqq-python3-3.10.8/lib/python3.10/unittest/loader.py", line 436, in _find_test_path
       >     module = self._get_module_from_name(name)
       >   File "/nix/store/xcaaly5shfy227ffs8nipxrd49b56iqq-python3-3.10.8/lib/python3.10/unittest/loader.py", line 377, in _get_module_from_name
       >     __import__(name)
       >   File "/build/python/tblite/test_library.py", line 18, in <module>
       >     from tblite.library import get_version
       >   File "/build/python/tblite/library.py", line 30, in <module>
       >     raise ImportError("tblite C extension unimportable, cannot use C-API")
       > ImportError: tblite C extension unimportable, cannot use C-API
```
which made me doubt their lib-export setup; will try again against your tree

**Comment by sheepforce** [comment]
> At least one test of pysis is failing now. I will collect them and figure out what to do :)

There are 4 mismatches between atoms and None type errors now in pysisyphus now. Thus, unfortunately not only numerical deviations. I will talk with @eljost what to do.
Here is the relevant part:
```
python3.10-pysisyphus> =================================== FAILURES ===================================
python3.10-pysisyphus> ___________________________ test_endopt_total[True] ____________________________
python3.10-pysisyphus> do_hess = True, this_dir = PosixPath('/build/source/tests/test_barriers')
python3.10-pysisyphus>     @using("xtb")
python3.10-pysisyphus>     @pytest.mark.parametrize(
python3.10-pysisyphus>         "do_hess",
python3.10-pysisyphus>         (
python3.10-pysisyphus>             False,
python3.10-pysisyphus>             True,
python3.10-pysisyphus>         ),
python3.10-pysisyphus>     )
python3.10-pysisyphus>     def test_endopt_total(do_hess, this_dir):
python3.10-pysisyphus>         # ed, ts, prod = geom_loader("lib:xtb_rx/07_dacp_eth.trj")
python3.10-pysisyphus>         # print(geoms)
python3.10-pysisyphus>         run_dict = {
python3.10-pysisyphus>             "geom": {"type": "cart", "fn": "lib:xtb_rx/07_dacp_eth.trj[1]"},
python3.10-pysisyphus>             "calc": {
python3.10-pysisyphus>                 "type": "xtb",
python3.10-pysisyphus>                 "pal": 6,
python3.10-pysisyphus>                 "quiet": True,
python3.10-pysisyphus>             },
python3.10-pysisyphus>             "irc": {
python3.10-pysisyphus>                 "type": "eulerpc",
python3.10-pysisyphus>                 "hessian_init": this_dir / "inp_hess_init_irc.h5",
python3.10-pysisyphus>             },
python3.10-pysisyphus>             "endopt": {
python3.10-pysisyphus>                 "fragments": "total",
python3.10-pysisyphus>                 "do_hess": do_hess,
python3.10-pysisyphus>             },
python3.10-pysisyphus>             "barriers": {
python3.10-pysisyphus>                 "solv_calc": {
python3.10-pysisyphus>                     "type": "xtb",
python3.10-pysisyphus>                     "pal": 6,
python3.10-pysisyphus>                     # "quiet": True,
python3.10-pysisyphus>                     "gbsa": "water",
python3.10-pysisyphus>                 },
python3.10-pysisyphus>                 "do_standard_state_corr": True,
python3.10-pysisyphus>             },
python3.10-pysisyphus>         }
python3.10-pysisyphus> >       results = run_from_dict(run_dict)
python3.10-pysisyphus> tests/test_barriers/test_barriers.py:47:
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> pysisyphus/run.py:1991: in run_from_dict
python3.10-pysisyphus>     run_result = main(run_dict, restart, cwd, scheduler)
python3.10-pysisyphus> pysisyphus/run.py:1639: in main
python3.10-pysisyphus>     do_endopt_ts_barriers(
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> ts_geom = Geometry(C₇H₁₀), left_geoms = [Geometry(C₅H₆)]
python3.10-pysisyphus> right_geoms = [Geometry(C₇H₁₀)]
python3.10-pysisyphus> left_fns = ['forward_end_frag01_final_geometry.xyz', 'forward_end_final_geometry.xyz']
python3.10-pysisyphus> right_fns = ['backward_end_frag00_final_geometry.xyz'], do_thermo = False
python3.10-pysisyphus> T = 298.15, p = 101325
python3.10-pysisyphus> calc_getter = <function get_calc_closure.<locals>.calc_getter at 0x7ffeebd025f0>
python3.10-pysisyphus> solv_calc_getter = <function get_calc_closure.<locals>.calc_getter at 0x7ffeebd024d0>
python3.10-pysisyphus> do_standard_state_corr = True
python3.10-pysisyphus>     def do_endopt_ts_barriers(
python3.10-pysisyphus>         ts_geom,
python3.10-pysisyphus>         left_geoms,
python3.10-pysisyphus>         right_geoms=None,
python3.10-pysisyphus>         left_fns=None,
python3.10-pysisyphus>         right_fns=None,
python3.10-pysisyphus>         do_thermo=False,
python3.10-pysisyphus>         T=T_DEFAULT,
python3.10-pysisyphus>         p=p_DEFAULT,
python3.10-pysisyphus>         calc_getter=None,
python3.10-pysisyphus>         solv_calc_getter=None,
python3.10-pysisyphus>         do_standard_state_corr=False,
python3.10-pysisyphus>     ):
python3.10-pysisyphus>         print(highlight_text("Barrier heights after end optimization(s)"))
python3.10-pysisyphus>         print()
python3.10-pysisyphus>         # Only allow standard state correction when solvent calculator is specified.
python3.10-pysisyphus>         do_ssc = do_standard_state_corr and (solv_calc_getter is not None)
python3.10-pysisyphus>         ssc = standard_state_corr(T=T, p=p) if do_ssc else 0.0
python3.10-pysisyphus>         if right_geoms is None:
python3.10-pysisyphus>             right_geoms = []
python3.10-pysisyphus>         def set_fns(fns, geoms):
python3.10-pysisyphus>             if fns is None:
python3.10-pysisyphus>                 fns = ["" for geom in geoms]
python3.10-pysisyphus>             else:
python3.10-pysisyphus>                 fns = [Path(fn).name for fn in fns]
python3.10-pysisyphus>             return fns
python3.10-pysisyphus>         left_fns = set_fns(left_fns, left_geoms)
python3.10-pysisyphus>         right_fns = set_fns(right_fns, left_geoms)
python3.10-pysisyphus>         ts_fn = "TS"
python3.10-pysisyphus>         ts_atom_num = len(ts_geom.atoms)
python3.10-pysisyphus>         def drop_total_geom(geoms):
python3.10-pysisyphus>             atom_nums = [len(geom.atoms) for geom in geoms]
python3.10-pysisyphus>             tot_atom_num = sum(atom_nums)
python3.10-pysisyphus>             total_geom = None
python3.10-pysisyphus>             if tot_atom_num > ts_atom_num:
python3.10-pysisyphus>                 total_ind = atom_nums.index(max(atom_nums))
python3.10-pysisyphus>                 total_geom = geoms[total_ind]
python3.10-pysisyphus>                 geoms = [geom for i, geom in enumerate(geoms) if i != total_ind]
python3.10-pysisyphus>             return geoms, total_geom
python3.10-pysisyphus>         left_geoms, left_total_geom = drop_total_geom(left_geoms)
python3.10-pysisyphus>         right_geoms, right_total_geom = drop_total_geom(right_geoms)
python3.10-pysisyphus>         def tot_atom_num(geoms):
python3.10-pysisyphus>             return sum([len(geom.atoms) for geom in geoms])
python3.10-pysisyphus>         left_atom_num = tot_atom_num(left_geoms)
python3.10-pysisyphus> >       assert left_atom_num == ts_atom_num, (
python3.10-pysisyphus>             f"Atom number mismatch between TS ({ts_atom_num} atoms) and "
python3.10-pysisyphus>             f"left side ({left_atom_num} atoms)! Aborting barrier calculation"
python3.10-pysisyphus>         )
python3.10-pysisyphus> E       AssertionError: Atom number mismatch between TS (17 atoms) and left side (11 atoms)! Aborting barrier calculation
python3.10-pysisyphus> pysisyphus/drivers/barriers.py:65: AssertionError
python3.10-pysisyphus> _________________________ test_gen_guess_hessians[xtb] _________________________
python3.10-pysisyphus> hessian_init = 'xtb'
python3.10-pysisyphus>     @using("pyscf")
python3.10-pysisyphus>     @pytest.mark.parametrize(
python3.10-pysisyphus>         "hessian_init",
python3.10-pysisyphus>         (
python3.10-pysisyphus>             "calc",
python3.10-pysisyphus>             "unit",
python3.10-pysisyphus>             "fischer",
python3.10-pysisyphus>             "lindh",
python3.10-pysisyphus>             "simple",
python3.10-pysisyphus>             "swart",
python3.10-pysisyphus>             pytest.param("xtb", marks=[using("pyscf"), using("xtb")]),
python3.10-pysisyphus>             pytest.param("xtb1", marks=[using("pyscf"), using("xtb")]),
python3.10-pysisyphus>         ),
python3.10-pysisyphus>     )
python3.10-pysisyphus>     def test_gen_guess_hessians(hessian_init):
python3.10-pysisyphus>         geom = geom_loader("lib:h2o2_hf_321g_opt.xyz", coord_type="redund")
python3.10-pysisyphus>         geom.set_calculator(PySCF(basis="def2svp", pal=1))
python3.10-pysisyphus>         opt = RFOptimizer(geom, hessian_init=hessian_init)
python3.10-pysisyphus> >       opt.prepare_opt()
python3.10-pysisyphus> tests/test_guess_hessians/test_guess_hessians.py:72:
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> pysisyphus/optimizers/HessianOptimizer.py:237: in prepare_opt
python3.10-pysisyphus>     self.H, hess_str = get_guess_hessian(self.geometry, hessian_init)
python3.10-pysisyphus> pysisyphus/optimizers/guess_hessians.py:259: in get_guess_hessian
python3.10-pysisyphus>     H, hess_str = hess_funcs[hessian_init]()
python3.10-pysisyphus> pysisyphus/optimizers/guess_hessians.py:252: in <lambda>
python3.10-pysisyphus>     "xtb": lambda: (xtb_hessian(geometry, gfn=2), "GFN2-XTB"),
python3.10-pysisyphus> pysisyphus/optimizers/guess_hessians.py:215: in xtb_hessian
python3.10-pysisyphus>     return geom_.hessian
python3.10-pysisyphus> pysisyphus/Geometry.py:1022: in hessian
python3.10-pysisyphus>     hessian = self.cart_hessian
python3.10-pysisyphus> pysisyphus/Geometry.py:999: in cart_hessian
python3.10-pysisyphus>     results = self.calculator.get_hessian(self.atoms, self._coords)
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:221: in get_hessian
python3.10-pysisyphus>     results = self.run(inp, **kwargs)
python3.10-pysisyphus> pysisyphus/calculators/Calculator.py:479: in run
python3.10-pysisyphus>     raise err
python3.10-pysisyphus> pysisyphus/calculators/Calculator.py:464: in run
python3.10-pysisyphus>     results = self.parser_funcs[calc](path, **parser_kwargs)
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:365: in parse_hessian
python3.10-pysisyphus>     energy = self.parse_energy(path)
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> self = <pysisyphus.calculators.XTB.XTB object at 0x7fffc8f1d990>
python3.10-pysisyphus> path = PosixPath('/build/calculator_000_000_n2qur3ub')
python3.10-pysisyphus>     def parse_energy(self, path):
python3.10-pysisyphus>         with open(path / self.out_fn) as handle:
python3.10-pysisyphus>             text = handle.read()
python3.10-pysisyphus>         energy_re = r"TOTAL ENERGY\s*([-\d\.]+) Eh"
python3.10-pysisyphus> >       energy = float(re.search(energy_re, text)[1])
python3.10-pysisyphus> E       TypeError: 'NoneType' object is not subscriptable
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:353: TypeError
python3.10-pysisyphus> ________________________ test_gen_guess_hessians[xtb1] _________________________
python3.10-pysisyphus> hessian_init = 'xtb1'
python3.10-pysisyphus>     @using("pyscf")
python3.10-pysisyphus>     @pytest.mark.parametrize(
python3.10-pysisyphus>         "hessian_init",
python3.10-pysisyphus>         (
python3.10-pysisyphus>             "calc",
python3.10-pysisyphus>             "unit",
python3.10-pysisyphus>             "fischer",
python3.10-pysisyphus>             "lindh",
python3.10-pysisyphus>             "simple",
python3.10-pysisyphus>             "swart",
python3.10-pysisyphus>             pytest.param("xtb", marks=[using("pyscf"), using("xtb")]),
python3.10-pysisyphus>             pytest.param("xtb1", marks=[using("pyscf"), using("xtb")]),
python3.10-pysisyphus>         ),
python3.10-pysisyphus>     )
python3.10-pysisyphus>     def test_gen_guess_hessians(hessian_init):
python3.10-pysisyphus>         geom = geom_loader("lib:h2o2_hf_321g_opt.xyz", coord_type="redund")
python3.10-pysisyphus>         geom.set_calculator(PySCF(basis="def2svp", pal=1))
python3.10-pysisyphus>         opt = RFOptimizer(geom, hessian_init=hessian_init)
python3.10-pysisyphus> >       opt.prepare_opt()
python3.10-pysisyphus> tests/test_guess_hessians/test_guess_hessians.py:72:
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> pysisyphus/optimizers/HessianOptimizer.py:237: in prepare_opt
python3.10-pysisyphus>     self.H, hess_str = get_guess_hessian(self.geometry, hessian_init)
python3.10-pysisyphus> pysisyphus/optimizers/guess_hessians.py:259: in get_guess_hessian
python3.10-pysisyphus>     H, hess_str = hess_funcs[hessian_init]()
python3.10-pysisyphus> pysisyphus/optimizers/guess_hessians.py:254: in <lambda>
python3.10-pysisyphus>     "xtb1": lambda: (xtb_hessian(geometry, gfn=1), "GFN1-XTB"),
python3.10-pysisyphus> pysisyphus/optimizers/guess_hessians.py:215: in xtb_hessian
python3.10-pysisyphus>     return geom_.hessian
python3.10-pysisyphus> pysisyphus/Geometry.py:1022: in hessian
python3.10-pysisyphus>     hessian = self.cart_hessian
python3.10-pysisyphus> pysisyphus/Geometry.py:999: in cart_hessian
python3.10-pysisyphus>     results = self.calculator.get_hessian(self.atoms, self._coords)
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:221: in get_hessian
python3.10-pysisyphus>     results = self.run(inp, **kwargs)
python3.10-pysisyphus> pysisyphus/calculators/Calculator.py:479: in run
python3.10-pysisyphus>     raise err
python3.10-pysisyphus> pysisyphus/calculators/Calculator.py:464: in run
python3.10-pysisyphus>     results = self.parser_funcs[calc](path, **parser_kwargs)
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:365: in parse_hessian
python3.10-pysisyphus>     energy = self.parse_energy(path)
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> self = <pysisyphus.calculators.XTB.XTB object at 0x7ffeeb789240>
python3.10-pysisyphus> path = PosixPath('/build/calculator_000_000_7v0v1sya')
python3.10-pysisyphus>     def parse_energy(self, path):
python3.10-pysisyphus>         with open(path / self.out_fn) as handle:
python3.10-pysisyphus>             text = handle.read()
python3.10-pysisyphus>         energy_re = r"TOTAL ENERGY\s*([-\d\.]+) Eh"
python3.10-pysisyphus> >       energy = float(re.search(energy_re, text)[1])
python3.10-pysisyphus> E       TypeError: 'NoneType' object is not subscriptable
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:353: TypeError
python3.10-pysisyphus> _______________________________ test_xtb_hessian _______________________________
python3.10-pysisyphus> geom = Geometry(HClO₃S)
python3.10-pysisyphus>     @using("xtb")
python3.10-pysisyphus>     def test_xtb_hessian(geom):
python3.10-pysisyphus> >       Hmw = geom.mw_hessian
python3.10-pysisyphus> tests/test_xtb/test_xtb.py:34:
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> pysisyphus/Geometry.py:1051: in mw_hessian
python3.10-pysisyphus>     return self.mass_weigh_hessian(self.cart_hessian)
python3.10-pysisyphus> pysisyphus/Geometry.py:999: in cart_hessian
python3.10-pysisyphus>     results = self.calculator.get_hessian(self.atoms, self._coords)
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:221: in get_hessian
python3.10-pysisyphus>     results = self.run(inp, **kwargs)
python3.10-pysisyphus> pysisyphus/calculators/Calculator.py:479: in run
python3.10-pysisyphus>     raise err
python3.10-pysisyphus> pysisyphus/calculators/Calculator.py:464: in run
python3.10-pysisyphus>     results = self.parser_funcs[calc](path, **parser_kwargs)
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:365: in parse_hessian
python3.10-pysisyphus>     energy = self.parse_energy(path)
python3.10-pysisyphus> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python3.10-pysisyphus> self = <pysisyphus.calculators.XTB.XTB object at 0x7ffe8ace90f0>
python3.10-pysisyphus> path = PosixPath('/build/calculator_000_000_j10b2p48')
python3.10-pysisyphus>     def parse_energy(self, path):
python3.10-pysisyphus>         with open(path / self.out_fn) as handle:
python3.10-pysisyphus>             text = handle.read()
python3.10-pysisyphus>         energy_re = r"TOTAL ENERGY\s*([-\d\.]+) Eh"
python3.10-pysisyphus> >       energy = float(re.search(energy_re, text)[1])
python3.10-pysisyphus> E       TypeError: 'NoneType' object is not subscriptable
python3.10-pysisyphus> pysisyphus/calculators/XTB.py:353: TypeError
```

@ryanswrt cool that you find it useful! :blush: I haven't even attempted to build the tblite python bindings, only xtb-python and this is already not the greatest experience. The previous version (before this MR) needed to copy shared object files manually into the proper places as the meson instructions were broken. 
If you can figure this out I would love to have this included.

**Comment by sheepforce** [comment]
@ryanswrt I was able to build the tblite python bindings. Seems to work from a python3 shell.

File: pythonPackages.nix [hunk 1]
```diff
@@ -53,6 +53,11 @@ let
 
     rmsd = callPackage ./pkgs/lib/rmsd { };
 
+    tblite = callPackage ./pkgs/lib/tblite/python.nix {
+      inherit (selfPkgs) tblite;
+      inherit (superPkgs) meson;
```
Comment (markuskowa) [inline pythonPackages.nix:58]: Does it pick up a python package with the same name if you remove that line?

**Comment by markuskowa** [comment]
pysisiphus fails on Hydra:
```
FAILED tests/test_barriers/test_barriers.py::test_endopt_total[True] - Assert...
FAILED tests/test_guess_hessians/test_guess_hessians.py::test_gen_guess_hessians[xtb]
FAILED tests/test_guess_hessians/test_guess_hessians.py::test_gen_guess_hessians[xtb1]
FAILED tests/test_xtb/test_xtb.py::test_xtb_hessian - TypeError: 'NoneType' o...
```

File: pythonPackages.nix (hunk 1 above)
Comment (sheepforce) [reply pythonPackages.nix:58]: Exactly, name [python3Packages.meson](https://search.nixos.org/packages?channel=22.11&from=0&size=50&sort=relevance&type=packages&query=python+meson)

**Comment by sheepforce** [comment]
> pysisiphus fails on Hydra:
> 
> ```
> FAILED tests/test_barriers/test_barriers.py::test_endopt_total[True] - Assert...
> FAILED tests/test_guess_hessians/test_guess_hessians.py::test_gen_guess_hessians[xtb]
> FAILED tests/test_guess_hessians/test_guess_hessians.py::test_gen_guess_hessians[xtb1]
> FAILED tests/test_xtb/test_xtb.py::test_xtb_hessian - TypeError: 'NoneType' o...
> ```

Yup, these are the exact four errors we also get. (See https://github.com/Nix-QChem/NixOS-QChem/pull/250#issuecomment-1425561848). Working on it ...

**Comment by sheepforce** [comment]
The failing tests are quite strange. They are **complete** numerical nonsense when run as is. Interestingly this does not seem to happen with the statically linked executable from the XTB repo. Now, even more interestingly, the problem only happens when `xtb` is called with `--acc` as pysisyphus does. It doesn't even matter which value it is given. The default is 1 and pysisyphus calls it with 1. However, having the parameter on XTB's command line will give nonsense either in the SCF already or latestly in the numerical Hessian calculation upon displacements. The calculation may arbitrarily hang, segfault, or complain about not being able to setup numerical thresholds.
I have tested whether it depends on the BLAS+Lapack implementation and OpenMP threads.
For BLAS, it does not matter. Neither generic BLAS, nor MKL nor OpenBLAS make any difference. Could also not observe anything reproducible for the number of threads. The same calculation may crash, hang or error with any number of threads.

Here is one of the examples that fails in pysisyphus:

```
xtb xtb.xyz --hess --input xcontrol --chrg 0 --uhf 0 --gfn 2 --acc 1.0
```

`xyb.xyz`:
```
4

  H -0.35738652 -2.42732105  0.01869782
  H -0.55575507  0.31681832  1.27512527
  O -0.71404094 -1.41851105  0.01869782
  O -0.20069873 -0.69255477  1.27610279
```

`xcontrol`:
```
$write
            json=true
        $end
```

It runs perfectly fine without `--acc`. The order of arguments does not matter, I've checked. XTB, tblite, ... pass all their tests, including those testing Hessians. I am clueless. Any ideas are appreciated. :grimacing:

**Comment by markuskowa** [comment]
Is it built with the `-ffast-math` compiler flag?

**Comment by sheepforce** [comment]
> Is it built with the `-ffast-math` compiler flag?

Hm, couldn't find anything like it in the cmake flags

**Comment by sheepforce** [comment]
Meh, here we go, even happens when completely following the toolchain of upstream XTB. That is, it even happens when building in a plain debian pod with fully embracing the cmake build system and building everything with buildtime downloads of deps. So
```bash
sudo apt install gfortran cmake build-essential git libopenblas-dev
git clone https://github.com/grimme-lab/xtb.git && cd xtb
cmake -B build -DCMAKE_BUILD_TYPE=Release
make -C build
make -C build test
```
In this case tblite also fails one test. Anyway, the behaviour is exactly the same. I would consider this an upstream bug. Should we open a bug report?

**Comment by markuskowa** [comment]
> I would consider this an upstream bug. Should we open a bug report?

That would be good.


> > Is it built with the `-ffast-math` compiler flag?
> 
> Hm, couldn't find anything like it in the cmake flags

It could potentially also be a dependency of XTB that is built with `-ffast-math` and is causing the problem.

**Comment by sheepforce** [comment]
> 



> > I would consider this an upstream bug. Should we open a bug report?
> 
> That would be good.

https://github.com/grimme-lab/xtb/issues/772

> > > Is it built with the `-ffast-math` compiler flag?
> > 
> > 
> > Hm, couldn't find anything like it in the cmake flags
> 
> It could potentially also be a dependency of XTB that is built with `-ffast-math` and is causing the problem.

I have checked in tblite, mctc-lib and mstore. There were none.

**Comment by sheepforce** [comment]
With the tooling for XTB in place, major parts of the DFTB+ infrastructure is also here. I've added those packages, also. This is unrelated to the numerical XTB problems, of course.

File: pkgs/apps/fypp/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  version = "3.1";
+
+  src = fetchFromGitHub {
+    owner = "aradi";
+    repo = pname;
+    rev = version;
+    hash = "sha256-iog5Gdcd1F230Nl4JDrKoyYr8JualVgNZQzHLzd4xe8=";
+  };
+
+  propagatedBuildInputs = [ ];
```
Comment (markuskowa) [inline pkgs/apps/fypp/default.nix:14]:
```suggestion
```

File: pkgs/lib/mpifx/default.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  };
+
+  nativeBuildInputs = [ cmake fypp gfortran ];
+
+  propagatedBuildInputs = [ mpi ];
+
+  #postInstall = ''
+  #  substituteInPlace $out/lib/pkgconfig/${pname}.pc \
+  #    --replace "''${prefix}" ""
+  #'';
```
Comment (markuskowa) [inline pkgs/lib/mpifx/default.nix:28]:
```suggestion
```
(assuming it is not needed anymore)

File: pkgs/lib/scalapackfx/default.nix [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  cmakeFlags = [
+    "-DSCALAPACK_LIBRARY=${scalapack}/lib/libscalapack.so"
+    "-DLAPACK_LIBRARY=${lapack}/lib/liblapack.so"
+  ];
+
+
+  #postInstall = ''
+  #  substituteInPlace $out/lib/pkgconfig/${pname}.pc \
+  #    --replace "''${prefix}" ""
+  #'';
```
Comment (markuskowa) [inline pkgs/lib/scalapackfx/default.nix:36]:
```suggestion
```

File: pkgs/lib/scalapackfx/default.nix [hunk 5]
```diff
@@ -0,0 +1,45 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, cmake
+, fypp
+, gfortran
+, mpifx
+, lapack
+, scalapack
+}:
```
Comment (markuskowa) [inline pkgs/lib/scalapackfx/default.nix:10]: There shoul be an ILP64 check here.

File: pkgs/lib/tblite/default.nix [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    export OMP_NUM_THREADS=2
+  '';
+
+  postInstall = ''
+    substituteInPlace $out/lib/pkgconfig/${pname}.pc \
+      --replace "''${prefix}" ""
+  '';
+
+  meta = with lib; {
+    description = " Light-weight tight-binding framework ";
```
Comment (markuskowa) [inline pkgs/lib/tblite/default.nix:51]:
```suggestion
    description = "Light-weight tight-binding framework";
```

**Comment by sheepforce** [comment]
I have removed the `--acc` flag from the pysis systemcall to xtb. Unfortunately we are hitting the next problem now. The tests still fail with the same errors, although I've confirmed that `--acc` is not present anymore. However, the very same test passes now outside of the build sandbox. I've checked the `env`s within the sandbox and outside. Nothing suspicious so far. Even playing around with nested `OMP_NUM_THREADS` does not make a difference. I am completely lost why xtb would behave differently in the sandbox and outside of it.

**Comment by sheepforce** [comment]
Even more: outside of the sandbox omitting the `--input xcontrol` makes the calculation hang, fail or segfault again, identical to what happens inside the sandbox. However, inside the sandbox the `--input xcontrol` has no effect on the bug.

**Comment by sheepforce** [comment]
For the moment I would assume that the XTB fix could take some time. The behaviour is rather strange, depends on fortran optimisation level, does not happen with IFort, is sensitive against parameters that have nothing to do with the actual calculation and so on.
As an alternative I am thinking about packaging the static XTB executable in addition and use this one for the moment, until the issue is resolved. This one is free of the described bugs.

On a sidenote: Do you know why the flang compiler is not part of the llvm packages in nixpkgs? It would be quite interesting to see if FLang would give an executable with correct behaviour.

**Comment by markuskowa** [comment]
> As an alternative I am thinking about packaging the static XTB executable in addition and use this one for the moment, until the issue is resolved. This one is free of the described bugs. What is you opinion on this?

We can wait a little for a proper fix?
Btw. It seems that on Hydra it passes with the pinned nixpkgs (i.e., with gfortran11/gcc11). Downgrading the compiler versions could be a solution?

> On a sidenote: Do you know why the flang compiler is not part of the llvm packages in nixpkgs? It would be quite interesting to see if FLang would give an executable with correct behaviour.

I would guess that Fortran is not popular enough in the Nix community. No just made the effort yet? Integrating compilers into the nixpkgs is no easy task.

**Comment by sheepforce** [comment]
OK, let's wait then until we know more.

> Btw. It seems that on Hydra it passes with the pinned nixpkgs (i.e., with gfortran11/gcc11). Downgrading the compiler versions could be a solution?
> EDIT: May not help. I looked at the wrong Hydra.

It does unfortunately not help. On Debian I've used Debian's gfortran 10.2.1 and the issue is the same.

> I would guess that Fortran is not popular enough in the Nix community. No one just made the effort yet? Integrating compilers into the nixpkgs is no easy task.

Hm, actually just found https://github.com/NixOS/nixpkgs/pull/163878
Has been approved but not merged yet :/

**Comment by sheepforce** [comment]
The bug has been fixed upstream and I could build pysisyphus without errors now. Should be ready :)

File: pkgs/apps/dftbplus/default.nix [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+, mpifx
+, scalapackfx
+, simple-dftd3
+, multicharge
+, dftd4
+, numpy
+}:
+
+assert blas.isILP64 == lapack.isILP64;
+
```
Comment (markuskowa) [inline pkgs/apps/dftbplus/default.nix:24]:
```suggestion
assert blas.isILP64 == scalapack.isILP64;
assert !blas.isILP64; 

```

File: pkgs/apps/xtb/default.nix [hunk 8]
```diff
@@ -1,9 +1,28 @@
...
+, multicharge
+, enableTurbomole ? false
+, turbomole
+, enableOrca ? false
+, orca
 , cefine
-} :
+}:
+
+assert blas.isILP64 == lapack.isILP64;
```
Comment (markuskowa) [inline pkgs/apps/xtb/default.nix:26]:
```suggestion
assert !blas.isILP64;

```

File: pkgs/lib/dftd4/default.nix [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+, gfortran
+, blas
+, lapack
+, mctc-lib
+, mstore
+, multicharge
+}:
+
+assert blas.isILP64 == lapack.isILP64;
+
```
Comment (markuskowa) [inline pkgs/lib/dftd4/default.nix:14]:
```suggestion
assert !blas.isILP64;

```

File: pkgs/lib/multicharge/default.nix [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+, cmake
+, gfortran
+, blas
+, lapack
+, mctc-lib
+, mstore
+}:
+
+assert blas.isILP64 == lapack.isILP64;
+
```
Comment (markuskowa) [inline pkgs/lib/multicharge/default.nix:13]:
```suggestion
assert !blas.isILP64;

```

File: pkgs/lib/scalapackfx/default.nix [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+assert !lapack.isILP64;
+
```
Comment (markuskowa) [inline pkgs/lib/scalapackfx/default.nix:13]:
```suggestion
assert lapack.isILP64 == scalapack.isILP64;

```

File: pkgs/lib/tblite/default.nix [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+, blas
+, lapack
+, mctc-lib
+, mstore
+, toml-f
+, multicharge
+, dftd4
+, simple-dftd3
+}:
+
```
Comment (markuskowa) [inline pkgs/lib/tblite/default.nix:15]:
```suggestion
assert blas.isILP64 == lapack.isILP64;
assert !blas.isILP64;

```

**Pull request merged**